semhash/semhash.py

```diff
@@ -309,133 +310,237 @@ def _validate_if_strings(self, records: Sequence[dict[str, str] | str]) -> Seque
+        :param lambda_param: Trade-off parameter between relevance and diversity.
+        :return: A FilterResult with the diversified representatives.
+        """
+        ranking = self._self_rank_by_average_similarity(descending=True)
+        return self._mmr(ranking, candidate_limit, selection_size, lambda_param)
 
-    def _filter_by_entropy(
+    def find_outliers(
         self,
-        records: Sequence[dict[str, str]],
-        vectors: np.ndarray,
-        budget: float | int,
-        descending: bool = True,
+        records: Sequence[Record],
+        outlier_percentage: float = 0.1,
     ) -> FilterResult:
         """
-        Filter records based on their entropy scores.
+        Find outliers in a given set of records against the fitted dataset.
 
-        :param records: The records to filter.
-        :param vectors: The vectors corresponding to the records.
-        :param budget: The maximum number of records to keep.
-        :param descending: Whether to sort in descending order of entropy.
-        :return: A FilterResult containing selected and filtered records.
+        This method ranks the records by their average similarity and selects the bottom
+        outlier_percentage of records as outliers.
+
+        :param records: A sequence of records to find outliers in.
+        :param outlier_percentage: The percentage (between 0 and 1) of records to consider outliers.
+        :return: A FilterResult where 'selected' contains the outliers and 'filtered' contains the inliers.
         """
-        budget = self._validate_filter_budget(budget=budget, n_records=len(records))
-
-        # compute entropy scores
-        scores = [
-            (idx, record, entropy_from_distances(results[0][-1]), results[0][0], results[0][-1])
-            for idx, (record, vectors) in enumerate(zip(records, vectors))
-            for results in [self.index.query_top_k(vectors.reshape(1, -1), k=100)]
-        ]
-
-        # sort scores
-        scores.sort(key=lambda x: x[2], reverse=descending)
-
-        # select records
-        selected: list[Union[dict[str, str], str]] = []
-        selected_indices: set[int] = set()
-        scores_selected: list[float] = []
-        for idx, record, entropy, _, _ in scores:
-            if len(selected) >= budget:
-                break
-            if idx not in selected_indices:
-                selected.append(record)
-                selected_indices.add(idx)
-                scores_selected.append(entropy)
-
-        # filter records
-        filtered = [record for idx, record, _, _, _ in scores if idx not in selected_indices]
-        scores_filtered: list[float] = [entropy for idx, record, entropy, _, _ in scores if idx not in selected_indices]
+        ranking = self._rank_by_average_similarity(records, descending=True)
+        outlier_count = int(len(ranking.selected) * outlier_percentage)
+        outlier_records = ranking.selected[-outlier_count:]
```

Comment:
subtle bug: if outlier_count is 0 due to rounding (i.e., ranking.selected * outlier_percentage < 1), you actually select all records, because `x[-0:]` is the whole list. Instead, you would expect 0 or 1 items to be selected. Use `ceil` instead of `int` to prevent.